kernel/src/net/socket/unix/ns/abs.rs

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{collections::btree_map::Entry, format};
+
+use keyable_arc::KeyableArc;
+
+use crate::prelude::*;
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct AbstractHandle(KeyableArc<[u8]>);
+
+impl AbstractHandle {
+    fn new(name: Arc<[u8]>) -> Self {
+        Self(KeyableArc::from(name))
+    }
+
+    pub fn name(&self) -> Arc<[u8]> {
+        self.0.clone().into()
+    }
+}
+
+impl Drop for AbstractHandle {
+    fn drop(&mut self) {
+        HANDLE_TABLE.remove(self.name());
+    }
+}
+
+static HANDLE_TABLE: HandleTable = HandleTable::new();
+
+struct HandleTable {
+    handles: RwLock<BTreeMap<Arc<[u8]>, Weak<AbstractHandle>>>,
+}
+
+impl HandleTable {
+    const fn new() -> Self {
+        Self {
+            handles: RwLock::new(BTreeMap::new()),
+        }
+    }
+
+    fn create(&self, name: Arc<[u8]>) -> Option<Arc<AbstractHandle>> {
+        let mut handles = self.handles.write();
+
+        let mut entry = handles.entry(name.clone());
+
+        if let Entry::Occupied(ref occupied) = entry {
```

Comment:
What will happen if there'a already an entry which has the same string content as `name` in the table? Should we allow two entries with the same string content in the table?

Reply:
I'm sorry, I don't understand this comment.

If we're trying to bind, the result should be `EADDRINUSE`, which is implemented in line 49+line 108.

Do you think this is wrong?

Reply:
You are right. I used to think that using `Arc<[u8]>` as key is same as `KeyableArc<[u8]>`, but after doing a simple test, I found they are actually different.

```rust
    let mut set = BTreeSet::new();
    let s1: Arc<[u8]> = vec![1, 2, 3].into();
    let s2: Arc<[u8]> = vec![1, 2, 3].into();
    set.insert(s1.clone());
    set.insert(s2.clone());
    println!("set = {:?}", set);

    let mut keyable_set = BTreeSet::new();
    let s1 = KeyableArc::from(s1);
    let s2 = KeyableArc::from(s2);
    keyable_set.insert(s1);
    keyable_set.insert(s2);
    println!("keyable_set = {:?}", keyable_set);
```

The result is
```text
set = {[1, 2, 3]}
keyable_set = {[1, 2, 3], [1, 2, 3]}
```

This is because `Arc<[u8]>` uses `[u8]` as the key, while `KeyableArc<[u8]>` uses the address of `Arc<[u8]>`. So your code is right.

Reply:
Yes. I think this is also why I need `HashMap<Arc<[u8]>, Weak<AbstractHandle>>` instead of `HashSet<KeyableWeak<AbstractHandle>>`. I need to find the handle by name, not by address.